### PR TITLE
Improve `TransferFunction` docs and add `to_expr`

### PR DESCRIPTION
**Pull request by akshanshbhatt**
#### Brief description of what is fixed or changed
* Updated the documentation of `TransferFunction`. Made it more beginner-friendly.
* Added a method `to_expr()` for conversion from `TransferFunction` type to `Expr` along with unit tests.
* Added a classmethod for instantiating `TransferFunction` objects from rational expressions (`from_rational_expression`).
#### Release Notes

* physics.control
  * `to_expr` method added in `TransferFunction` for converting `TransferFunction` to sympy expr.
  * `from_rational_expression` classmethod added for creating `TransferFunction` objects directly from rational expressions.

### CODE REVIEW

**Comment by moorepants** [comment]
Almost everything here is a backwards incompatible change. I think TransferFunction is already in a number of sympy releases. I recommend not changing any of the errors. As for expand, I'd leave the method but make it return the already expanded form.

You don't have an explanation as to why expand should be changed in this way. What is the use case? Why should the expressions be expanded automatically? It isn't really a sympy idiom to do so.

Lastly, I didn't see any tests for the new _to_expr method.

File: sympy/physics/control/tests/test_lti.py
Comment (moorepants) [inline sympy/physics/control/tests/test_lti.py:260]: I missed these on first pass. Scratch my other comment.

Comment (moorepants) [reply sympy/physics/control/tests/test_lti.py:260]: What is the purpose of simplify and sympify here? You can use `S(0)` to get a SymPy zero if needed.

**Comment by akshanshbhatt** [comment]
> You don't have an explanation as to why expand should be changed in this way. What is the use case? Why should the expressions be expanded automatically? It isn't really a sympy idiom to do so.

@namannimmo10 and I discussed implementing `expand()` by default (https://github.com/sympy/sympy/pull/21556#discussion_r650993372). I provided some use cases and practices used in other popular control packages to make the form of `TransferFunction` more coherent. `expand()` by default was also approved by you in the call we had. I have implemented it after getting approval from both Naman and you.

File: sympy/physics/control/lti.py
Comment (namannimmo10) [inline sympy/physics/control/lti.py:433]:
```suggestion
            raise ValueError("{} cannot be subtracted from a TransferFunction."
```
This should remain a `ValueError`, no? `TypeError` is raised when an operation or a function is applied to an object of inappropriate type.

File: sympy/physics/control/lti.py
Comment (namannimmo10) [inline sympy/physics/control/lti.py:34]:
```suggestion
    Generally, a dynamical system representing a physical model can be described in terms of Linear
```

File: sympy/physics/control/lti.py
Comment (namannimmo10) [inline sympy/physics/control/lti.py:78]: Can you add one line above describing what is $s$?

File: sympy/physics/control/lti.py
Comment (namannimmo10) [inline sympy/physics/control/lti.py:568]: Make this one line?

Comment (namannimmo10) [reply sympy/physics/control/lti.py:568]: Why private though? This could be a way for users to convert a transfer function to an elementary function.

Comment (akshanshbhatt) [reply sympy/physics/control/lti.py]: > Why private though? This could be a way for users to convert a transfer function to an elementary function.

Since the only possible use case, in my view at the time of its implementation, was the internal conversion of `TransferFunction` objects to `Expr` to make them compatible with the native matrix operations (since non-expr operations were deprecated).
If there are other possible use cases that a user might find helpful, I will make it a public method. I will do whatever you suggest.

File: sympy/physics/control/lti.py
Comment (akshanshbhatt) [reply sympy/physics/control/lti.py:436]: I think it should be `TypeError`. Here the exception is raised when the type of object we are subtracting from the `TransferFunction` is incompatible. The placeholder, too (in the print expression), logs the type of the object passed. Say I try to subtract a `Symbol` object from a `TransferFunction`, it will prompt an error -

```python
>>> from sympy.physics.control.lti import *
>>> from sympy.abc import *
>>> TransferFunction(1, s, s)
TransferFunction(1, s, s)
>>> _ - a
Traceback (most recent call last):   
  File "<stdin>", line 1, in <module>
  File "d:\sympy\sympy\physics\control\lti.py", line 436, in __sub__
    raise TypeError("{} cannot be subtracted from a TransferFunction."
TypeError: <class 'sympy.core.symbol.Symbol'> cannot be subtracted from a TransferFunction.
```
It makes sense as the type of object passed is incompatible for tf subtraction. Had this exception been raised when both the params were of the same type, there would have been a possibility for `ValueError`.
I used this as the reference - https://docs.python.org/3/library/exceptions.html#TypeError.

I was anyways, planning to revert to the old error types as suggested by @moorepants since this module has already gone through two release cycles. What is your opinion?

**Comment by moorepants** [comment]
> @namannimmo10 and I discussed implementing expand() by default (#21556 (comment)). I provided some use cases and practices used in other popular control packages to make the form of TransferFunction more coherent. expand() by default was also approved by you in the call we had. I have implemented it after getting approval from both Naman and you.

Yes, it sounded ok in the call, but I still don't know why it is being done. The explanation should be at least in the commit message here. Or show some examples that give an idea why this change is needed. There needs to be a justification for breaking backwards compatibility and not following typical sympy design patterns.

File: sympy/physics/control/lti.py
Comment (moorepants) [reply sympy/physics/control/lti.py:436]: The errors shouldn't be changed due to backwards incompatibility. It's more trouble for end users to break their code than worring about if the correct error types were originally selected.

Comment (akshanshbhatt) [reply sympy/physics/control/lti.py:436]: > The errors shouldn't be changed due to backwards incompatibility. It's more trouble for end users to break their code than worring about if the correct error types were originally selected.

I will revert the error-type changes.

File: sympy/physics/control/lti.py
Comment (namannimmo10) [inline sympy/physics/control/lti.py:80]:
```suggestion
    $s$, also known as complex frequency, is a complex variable in the Laplace domain. It corresponds to the
    equivalent variable $t$, in the time domain.
```
This looks sufficient. Since we are explaining this so well, it's also worth mentioning that a transfer function is the Laplace transform of the system's impulse response.

Comment (akshanshbhatt) [reply sympy/physics/control/lti.py:80]: > Since we are explaining this so well, it's also worth mentioning that a transfer function is the Laplace transform of the system's impulse response.

Where exactly should I mention this?

**Comment by oscarbenjamin** [comment]
Are the numerator and denominator always polynomials? If so then they could use Poly which expands automatically. Poly is more efficient than Expr for calculations with expanded polynomials. The internal representation of Poly is just a list of coefficients:
```python
In [3]: Poly(3*s**3 + s + 1).rep.rep
Out[3]: [mpz(3), mpz(0), mpz(1), mpz(1)]
```

I can imagine that it might be more useful in some cases to represent the system in terms of its poles and zeros rather than the expanded polynomial though.

In general I think it would be good to have a rational function class analogous to Poly although maybe it would have automatic cancellation and so not quite fit with the usage here.

**Comment by namannimmo10** [comment]
> Are the numerator and denominator always polynomials?

Actually, no. Last year we decided to support any arbitrary expressions in the numerator and denominator because we can have a [variety of transfer functions](https://tutorial.math.lamar.edu/classes/de/laplace_table.aspx) that aren't the ratio of pure polynomials. So it was not the right choice to use Poly internally.

> I can imagine that it might be more useful in some cases to represent the system in terms of its poles and zeros rather than the expanded polynomial though.

Yes, that was another main reason why I didn't use `expand` by default.
```python
In [3]: num, den = b, (s - a)**2 - b**2

In [4]: pprint(TransferFunction(num, den, s))
       b        
────────────────
   2           2
- b  + (-a + s) 

In [5]: pprint(TransferFunction(num, den, s).expand())
         b          
────────────────────
 2            2    2
a  - 2⋅a⋅s - b  + s 
```
Not sure if using `expand` does anything useful here.

```python
In [9]: _num, _den = factorial(n), (s - a)**(n + 1)

In [10]: pprint(TransferFunction(_num, _den, s))
      n!     
─────────────
        n + 1
(-a + s)     

In [11]: pprint(TransferFunction(_num, _den, s).expand())
             n!            
───────────────────────────
            n             n
- a⋅(-a + s)  + s⋅(-a + s) 

```

But now, even if we decide to expand the numerator and denominator by default, the user can call `factor` to get back to the unexpanded form. Maybe two examples in the docs could be added. 
```python
In [31]: G = TransferFunction(2*a*s**2, (s**2 + a**2)**2, s); pprint(G)
       2  
  2⋅a⋅s   
──────────
         2
⎛ 2    2⎞ 
⎝a  + s ⎠ 

In [32]: _G = TransferFunction(2*a*s**2, (s**2 + a**2)**2, s).expand(); pprint(_G)
           2     
      2⋅a⋅s      
─────────────────
 4      2  2    4
a  + 2⋅a ⋅s  + s 

In [33]: pprint(factor(_G))
       2  
  2⋅a⋅s   
──────────
         2
⎛ 2    2⎞ 
⎝a  + s ⎠ 
```
I would personally stick to the default unexpanded form.

**Comment by moorepants** [comment]
> I would personally stick to the default unexpanded form.

@namannimmo10 why were you suggesting in the meeting to do otherwise?

**Comment by namannimmo10** [comment]
https://github.com/sympy/sympy/pull/21556#discussion_r650887855 convinced me back then. :) Sorry, Akshansh.

**Comment by akshanshbhatt** [comment]
> [#21556 (comment)](https://github.com/sympy/sympy/pull/21556#discussion_r650887855) convinced me back then. :) Sorry, Akshansh.

Well, nothing to be sorry about here. Things that look good in the planning/prototyping period might not work as planned after implementation. Many latent requirements need to be satisfied, which usually come into perspective after the execution. I will change it back to the unexpanded form.

File: sympy/physics/control/lti.py
Comment (namannimmo10) [reply sympy/physics/control/lti.py]:
```python
>>> tf = TransferFunction(s + 1, s**2 + 2, s)
>>> delay = exp(-s/tau)
>>> expr = delay*tf.to_expr()
>>> H1 = TransferFunction.from_rational_expression(expr, s)
>>> 
>>> H2 = TransferFunction(exp(-s/tau)*(s + 1), s**2 + 2, s)
```

Both `H1` and `H2` should work, I guess.

**Comment by moorepants** [comment]
> #21556 (comment) convinced me back then. :) Sorry, Akshansh.

The comment has some good points. In the numeric libraries you pass in coefficients to the polynomial. I'm pretty sure we discussed this when Naman was designing things too and we felt that since we have a symbolic library that method would be too limiting (can't remember the details though). If expand() is expensive to call, then maybe it is worth doing it once on object instantiation. If it isn't expensive then you can get the coefficients whenever you need to with the appropriate function calls. We need a demonstration of why expand() should be called on instantiation, otherwise we shouldn't change it. The only reason presented in the comment is that other libraries in other languages do it differently. But why do they do it differently? What is the benefit?

@akshanshbhatt, you should focus on building on what we have, versus changing what we have already made. If, when building the new things you really need to change TransferFunction, then present the case, explain why and maybe it's worth deprecating things.

**Comment by moorepants** [comment]
You wrote this:

> In my opinion, having a definite form would enable us to compare two TransferFunction instances in later phases.

What are examples of comparing two transfer functions where this is needed?

**Comment by akshanshbhatt** [comment]
> The comment has some good points. In the numeric libraries you pass in coefficients to the polynomial. I'm pretty sure we discussed this when Naman was designing things too and we felt that since we have a symbolic library that method would be too limiting (can't remember the details though). If expand() is expensive to call, then maybe it is worth doing it once on object instantiation. If it isn't expensive then you can get the coefficients whenever you need to with the appropriate function calls. We need a demonstration of why expand() should be called on instantiation, otherwise we shouldn't change it. The only reason presented in the comment is that other libraries in other languages do it differently. But why do they do it differently? What is the benefit?

The purpose was a getting a uniform structure for `TransferFunction` objects, a sort of one-to-one mapping of TransferFunctions no matter the form. Consider these three transfer function objects, 

```
tf1 = TransferFunction(1, s**2 + 2*s + 1, s)
tf2 = TransferFunction(1, (s + 1)**2, s)
tf3 = TransferFunction(1, -s + (s + 1)*(s + 2) - 1, s)
```
Ideally, all these should be equal since all of their properties and attributes are equal (poles, zeros, stability) from the design principle. Yet, they are considered different under our model. 

```python
>>> Parallel(tf1, tf2) - tf3
Parallel(TransferFunction(1, s**2 + 2*s + 1, s), TransferFunction(1, (s + 1)**2, s), TransferFunction(-2, -s + (s + 1)*(s + 2) - 1, s))
>>> _.doit()
TransferFunction(-2*(s + 1)**2*(s**2 + 2*s + 1) + (-s + (s + 1)*(s + 2) - 1)*(s**2 + 2*s + (s + 1)**2 + 1), (s + 1)**2*(-s + (s + 1)*(s + 2) - 1)*(s**2 + 2*s + 1), s)
```

Even after using expand at this stage, the result is not accurate.

```python
>>> _.expand()
TransferFunction(0, s**6 + 6*s**5 + 15*s**4 + 20*s**3 + 15*s**2 + 6*s + 1, s)

# expected
TransferFunction(0, s**2 + 2*s +1, s)
```
Yes, we can use simplify here, but that too would be a resource-consuming choice (in fact, more than `expand()` and also not guaranteed to work).
This was specifically a bad sign since, 

* In `TrasferFunctionMatrix` there are many operations where the same transfer function instances are added and subtracted. Still, since their form is often different (like shown in the example above), it would output something totally gibberish when it should have outputted a simple answer.
* In solving textbook problems, most of the time, it is the case that there is some canceling out of common terms, which ultimately results in simple final answers. Here it won't be happening.

File: sympy/physics/control/lti.py
Comment (akshanshbhatt) [reply sympy/physics/control/lti.py]: > Both `H1` and `H2` should work, I guess.

This example looks fine but in the call which we had, @moorepants suggested not to implement `from_rational_expression()` classmethod.

**Comment by moorepants** [comment]
This is primarily the difference in symoblics and numerics. SymPy, by design, stores symbolics in verbose complex expressions. Symbolic equality is ill-defined mathematically and simplification oriented routines are not typically used in library algorithms, but more as a user tool. The polynomials package has objects for polynomials, but we decided not to use it inside TransferFunctions (for reasons I can't remember). The examples you show are really just simplification to show equality. There isn't anything wrong with it. That's just how sympy works.

**Comment by moorepants** [comment]
> In solving textbook problems, most of the time, it is the case that there is some canceling out of common terms, which ultimately results in simple final answers. Here it won't be happening.

In you examples, use simplification at the end to get a simple form if needed. This is what users will do.

**Comment by akshanshbhatt** [comment]
> This is primarily the difference in symoblics and numerics. SymPy, by design, stores symbolics in verbose complex expressions. Symbolic equality is ill-defined mathematically and simplification oriented routines are not typically used in library algorithms, but more as a user tool. The polynomials package has objects for polynomials, but we decided not to use it inside TransferFunctions (for reasons I can't remember). The examples you show are really just simplification to show equality. There isn't anything wrong with it. That's just how sympy works.

Yes, I understand it now, that's why I have removed it. What I have shared with you was my initial thought, the reason why I brought up expanding the terms by default in the first place.

File: sympy/physics/control/lti.py
Comment (namannimmo10) [reply sympy/physics/control/lti.py]: > ```python
> >>> tf = TransferFunction(s + 1, s**2 + 2, s)
> >>> delay = exp(-s/tau)
> >>> expr = delay*tf.to_expr()
> >>> H1 = TransferFunction.from_rational_expression(expr, s)
> >>> 
> >>> H2 = TransferFunction(exp(-s/tau)*(s + 1), s**2 + 2, s)
> ```

I think it's alright to go with that. I see similar classmethods being used in the codebase. 

There's a `to_expr` method in _holonomic.py_ that converts a holonomic function back to elementary function:
https://github.com/sympy/sympy/blob/8b783d15caae900e05c9236bf001fde908befe7a/sympy/holonomic/holonomic.py#L2066

It would be nice if ours did too. This PR is almost ready after that.

Comment (akshanshbhatt) [reply sympy/physics/control/lti.py]: I will complete the docstrings and tests for `from_rational_expression` by today. Hopefully, we can merge this PR after that

File: sympy/physics/control/lti.py
Comment (namannimmo10) [inline sympy/physics/control/lti.py:81]:
```suggestion
    The numerator of the transfer function is, therefore, the Laplace transform of the output signal
    (The signals are represented as functions of time) and similarly, the denominator
    of the transfer function is the Laplace transform of the input signal. It is also a convention
    to denote the input and output signal's Laplace transform with capital alphabets like shown below.

            $H(s) = \frac{Y(s)}{X(s)} = \frac{ \mathcal{L}\left\{y(t)\right\} }{ \mathcal{L}\left\{x(t)\right\} }$

    $s$, also known as complex frequency, is a complex variable in the Laplace domain. It corresponds to the
    equivalent variable $t$, in the time domain. Transfer functions are sometimes also referred to as the Laplace
    transform of the system's impulse response. Transfer function, $H$, is represented as a rational
    function in $s$ like,
```

File: sympy/physics/control/lti.py
Comment (namannimmo10) [inline sympy/physics/control/lti.py:233]:
```suggestion
        Creates a new ``TransferFunction`` efficiently from a rational expression.
```

File: sympy/physics/control/lti.py
Comment (namannimmo10) [inline sympy/physics/control/lti.py:250]:
```suggestion
            When ``expr`` has more than one variables and an optional parameter
```

**Review by namannimmo10 — COMMENTED** [review]
Minor changes required...but overall looks good.

File: sympy/physics/control/tests/test_lti.py
Comment (akshanshbhatt) [reply sympy/physics/control/tests/test_lti.py:108]: Delay is not supported as of now. 

It prompts -

```
TypeError: Unsupported type for numerator or denominator of TransferFunction.
```

It is due to `not num.has(exp)` condition. Should I remove it?

https://github.com/sympy/sympy/blob/8b783d15caae900e05c9236bf001fde908befe7a/sympy/physics/control/lti.py#L155-L158

Comment (namannimmo10) [reply sympy/physics/control/tests/test_lti.py:108]: Yes. Time delay is needed in several LTI models: https://www.mathworks.com/help/control/ug/time-delays-in-linear-systems.html

File: sympy/physics/control/lti.py
Comment (namannimmo10) [reply sympy/physics/control/lti.py:237]: Add one example in the docstring for `expr: Number`?

**Review by namannimmo10 — APPROVED** [review]
Looks ready! @moorepants, you have any other comments?

**Comment by namannimmo10** [comment]
I will merge this later tonight. Thanks for your contributions, @akshanshbhatt!

**Pull request merged**
